File: dqops/src/main/java/com/dqops/connectors/db2/Db2ParametersSpec.java
package com.dqops.connectors.db2;

import com.dqops.connectors.ConnectionProviderSpecificParameters;
import com.dqops.core.secrets.SecretValueLookupContext;
import com.dqops.core.secrets.SecretValueProvider;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMap;
import com.dqops.metadata.id.ChildHierarchyNodeFieldMapImpl;
import com.dqops.metadata.sources.BaseProviderParametersSpec;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.annotation.JsonNaming;
import lombok.EqualsAndHashCode;
import picocli.CommandLine;

import java.util.Collections;
import java.util.Map;
import java.util.Objects;

/**
 * IBM DB2 connection parameters.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
@JsonNaming(PropertyNamingStrategies.SnakeCaseStrategy.class)
@EqualsAndHashCode(callSuper = true)
public class Db2ParametersSpec extends BaseProviderParametersSpec
        implements ConnectionProviderSpecificParameters {

    private static final ChildHierarchyNodeFieldMapImpl<Db2ParametersSpec> FIELDS = new ChildHierarchyNodeFieldMapImpl<>(BaseProviderParametersSpec.FIELDS) {
        {
        }
    };

    @CommandLine.Option(names = {"--db2-platform"}, description = "DB2 platform type.")
    @JsonPropertyDescription("DB2 platform type. Supports also a ${DB2_PLATFORM} configuration with a custom environment variable.")
    private Db2PlatformType db2PlatformType;

    @CommandLine.Option(names = {"--db2-host"}, description = "DB2 host name")
    @JsonPropertyDescription("DB2 host name. Supports also a ${DB2_HOST} configuration with a custom environment variable.")
    private String host;

    @CommandLine.Option(names = {"--db2-port"}, description = "DB2 port number")
    @JsonPropertyDescription("DB2 port number. The default port is 50000. Supports also a ${DB2_PORT} configuration with a custom environment variable.")
    private String port;

    @CommandLine.Option(names = {"--db2-database"}, description = "DB2 database name")
    @JsonPropertyDescription("DB2 database name. Supports also a ${DB2_DATABASE} configuration with a custom environment variable.")
    private String database;

    @CommandLine.Option(names = {"--db2-user"}, description = "DB2 user name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    @JsonPropertyDescription("DB2 user name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    private String user;

    @CommandLine.Option(names = {"--db2-password"}, description = "DB2 database password. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    @JsonPropertyDescription("DB2 database password. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use dynamic substitution.")
    private String password;

    @CommandLine.Option(names = {"-DB2"}, description = "DB2 additional properties that are added to the JDBC connection string")
    @JsonPropertyDescription("A dictionary of custom JDBC parameters that are added to the JDBC connection string, a key/value dictionary.")
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    private Map<String, String> properties;

    /**
     * Returns the DB2 platform type.
     * @return DB2 platform type.
     */
    public Db2PlatformType getDb2PlatformType() {
        return db2PlatformType;
    }

    /**
     * Sets the DB2 platform type.
     * @param db2PlatformType New DB2 platform type.
     */
    public void setDb2PlatformType(Db2PlatformType db2PlatformType) {
        setDirtyIf(!Objects.equals(this.db2PlatformType, db2PlatformType));
        this.db2PlatformType = db2PlatformType;
    }

    /**
     * Returns the host name.
     * @return Host name.
     */
    public String getHost() {
        return host;
    }

    /**
     * Sets the host name.
     * @param host New host name.
     */
    public void setHost(String host) {
        setDirtyIf(!Objects.equals(this.host, host));
        this.host = host;
    }

    /**
     * Returns the port number. The value should store an environment variable expression or a numeric postgresql port number.
     * @return Port name or an expression to be extracted.
     */
    public String getPort() {
        return port;
    }

    /**
     * Sets the port number.
     * @param port Port name.
     */
    public void setPort(String port) {
        setDirtyIf(!Objects.equals(this.port, port));
        this.port = port;
    }

    /**
     * Returns the database name.
     * @return Database name.
     */
    public String gatDatabase() {
        return database;
    }

    /**
     * Sets the database name.
     * @param database New database name.
     */
    public void setDatabase(String database) {
        setDirtyIf(!Objects.equals(this.database, database));
        this.database = database;
    }

    /**
     * Returns the user that is used to log in to the data source (JDBC user or similar).
     * @return User name.
     */
    public String getUser() {
        return user;
    }

    /**
     * Sets a user name.
     * @param user User name.
     */
    public void setUser(String user) {
        setDirtyIf(!Objects.equals(this.user, user));
        this.user = user;
    }

    /**
     * Returns a password used to authenticate to the server.
     * @return Password.
     */
    public String getPassword() {
        return password;
    }

    /**
     * Sets a password that is used to connect to the database.
     * @param password Password.
     */
    public void setPassword(String password) {
        setDirtyIf(!Objects.equals(this.password, password));
        this.password = password;
    }

    /**
     * Returns a key/value map of additional properties that are included in the JDBC connection string.
     * @return Key/value dictionary of additional JDBC properties.
     */
    public Map<String, String> getProperties() {
        return properties;
    }

    /**
     * Sets a dictionary of additional connection parameters that are added to the JDBC connection string.
     * @param properties Key/value dictionary with extra parameters.
     */
    public void setProperties(Map<String, String> properties) {
        setDirtyIf(!Objects.equals(this.properties, properties));
        this.properties = properties != null ? Collections.unmodifiableMap(properties) : null;
    }

    /**
     * Returns the child map on the spec class with all fields.
     *
     * @return Return the field map.
     */
    @Override
    protected ChildHierarchyNodeFieldMap getChildMap() {
        return FIELDS;
    }

    /**
     * Creates and returns a deep copy of this object.
     */
    @Override
    public Db2ParametersSpec deepClone() {
        Db2ParametersSpec cloned = (Db2ParametersSpec)super.deepClone();
        return cloned;
    }

    /**
     * Creates a trimmed and expanded version of the object without unwanted properties, but with all variables like ${ENV_VAR} expanded.
     * @param secretValueProvider Secret value provider.
     * @param lookupContext Secret lookup context.
     * @return Trimmed and expanded version of this object.
     */
    public Db2ParametersSpec expandAndTrim(SecretValueProvider secretValueProvider, SecretValueLookupContext lookupContext) {
        Db2ParametersSpec cloned = this.deepClone();
        cloned.host = secretValueProvider.expandValue(cloned.host, lookupContext);
        cloned.port = secretValueProvider.expandValue(cloned.port, lookupContext);
        cloned.user = secretValueProvider.expandValue(cloned.user, lookupContext);
        cloned.password = secretValueProvider.expandValue(cloned.password, lookupContext);
        cloned.properties = secretValueProvider.expandProperties(cloned.properties, lookupContext);

        return cloned;
    }

    /**
     * Returns a database name.
     * @return Database name.
     */
    @Override
    public String getDatabase(){
        return database;
    }

}

File: distribution/python/dqops/client/models/db_2_parameters_spec.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.db_2_platform_type import Db2PlatformType
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.db_2_parameters_spec_properties import Db2ParametersSpecProperties


T = TypeVar("T", bound="Db2ParametersSpec")


@_attrs_define
class Db2ParametersSpec:
    """
    Attributes:
        db2_platform_type (Union[Unset, Db2PlatformType]):
        host (Union[Unset, str]): DB2 host name. Supports also a ${DB2_HOST} configuration with a custom environment
            variable.
        port (Union[Unset, str]): DB2 port number. The default port is 50000. Supports also a ${DB2_PORT} configuration
            with a custom environment variable.
        database (Union[Unset, str]): DB2 database name. Supports also a ${DB2_DATABASE} configuration with a custom
            environment variable.
        user (Union[Unset, str]): DB2 user name. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format to use
            dynamic substitution.
        password (Union[Unset, str]): DB2 database password. The value can be in the ${ENVIRONMENT_VARIABLE_NAME} format
            to use dynamic substitution.
        properties (Union[Unset, Db2ParametersSpecProperties]): A dictionary of custom JDBC parameters that are added to
            the JDBC connection string, a key/value dictionary.
    """

    db2_platform_type: Union[Unset, Db2PlatformType] = UNSET
    host: Union[Unset, str] = UNSET
    port: Union[Unset, str] = UNSET
    database: Union[Unset, str] = UNSET
    user: Union[Unset, str] = UNSET
    password: Union[Unset, str] = UNSET
    properties: Union[Unset, "Db2ParametersSpecProperties"] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        db2_platform_type: Union[Unset, str] = UNSET
        if not isinstance(self.db2_platform_type, Unset):
            db2_platform_type = self.db2_platform_type.value

        host = self.host
        port = self.port
        database = self.database
        user = self.user
        password = self.password
        properties: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.properties, Unset):
            properties = self.properties.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if db2_platform_type is not UNSET:
            field_dict["db2_platform_type"] = db2_platform_type
        if host is not UNSET:
            field_dict["host"] = host
        if port is not UNSET:
            field_dict["port"] = port
        if database is not UNSET:
            field_dict["database"] = database
        if user is not UNSET:
            field_dict["user"] = user
        if password is not UNSET:
            field_dict["password"] = password
        if properties is not UNSET:
            field_dict["properties"] = properties

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.db_2_parameters_spec_properties import Db2ParametersSpecProperties

        d = src_dict.copy()
        _db2_platform_type = d.pop("db2_platform_type", UNSET)
        db2_platform_type: Union[Unset, Db2PlatformType]
        if isinstance(_db2_platform_type, Unset):
            db2_platform_type = UNSET
        else:
            db2_platform_type = Db2PlatformType(_db2_platform_type)

        host = d.pop("host", UNSET)

        port = d.pop("port", UNSET)

        database = d.pop("database", UNSET)

        user = d.pop("user", UNSET)

        password = d.pop("password", UNSET)

        _properties = d.pop("properties", UNSET)
        properties: Union[Unset, Db2ParametersSpecProperties]
        if isinstance(_properties, Unset):
            properties = UNSET
        else:
            properties = Db2ParametersSpecProperties.from_dict(_properties)

        db_2_parameters_spec = cls(
            db2_platform_type=db2_platform_type,
            host=host,
            port=port,
            database=database,
            user=user,
            password=password,
            properties=properties,
        )

        db_2_parameters_spec.additional_properties = d
        return db_2_parameters_spec

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
